src/wasm2js.h

```diff
@@ -541,51 +553,45 @@ void Wasm2JSBuilder::addGlobalImport(Ref ast, Global* import) {
 }
 
 void Wasm2JSBuilder::addTable(Ref ast, Module* wasm) {
-  // Emit a simple flat table as a JS array literal. Otherwise,
-  // emit assignments separately for each index.
-  TableUtils::FlatTable flat(wasm->table);
-  if (flat.valid && !wasm->table.imported()) {
+  if (!wasm->table.exists)
+    return;
+
+  if (isTableExported(*wasm)) {
+    // If the table is exported use a fake WebAssembly.Table object
+    Ref theVar = ValueBuilder::makeVar();
+    ast->push_back(theVar);
+    Ref table = ValueBuilder::makeNew(ValueBuilder::makeCall(
+      IString("FakeTable"),
+      ValueBuilder::makeInt(Address::address32_t(wasm->table.initial.addr))));
+    ValueBuilder::appendToVar(theVar, FUNCTION_TABLE, table);
+  } else if (!wasm->table.imported()) {
+    // Otherwise if the table is interal (not imported eithe, just use
+    // a plain array.
     Ref theVar = ValueBuilder::makeVar();
     ast->push_back(theVar);
     Ref theArray = ValueBuilder::makeArray();
     ValueBuilder::appendToVar(theVar, FUNCTION_TABLE, theArray);
-    Name null("null");
-    for (auto& name : flat.names) {
-      if (name.is()) {
-        name = fromName(name, NameScope::Top);
+  }
```

Comment:
maybe add an assert that we don't handle the case of the table being both import *and* exported?

Reply:
Done